design/batching.md

```diff
@@ -0,0 +1,944 @@
+    // that implementation.
+    name: ComponentName,
+
+    /// A uniformly typed list of values for the given component type.
+    ///
+    /// Includes the data, its schema and probably soon the component metadata (e.g. the `ComponentName`).
+    values: Box<dyn arrow2::Array>,
+}
+
+impl DataCell {
+    /// Builds a new `DataCell` out of a uniform list of native component values.
+    pub fn from_native<C: Component>(values: Vec<C>) -> Self { /* ... */ }
+
+    /// Builds a new `DataCell` from an arrow array.
+    //
+    // TODO(cmc): We shouldn't have to specify the component name separately, this should be part of the
+    // metadata by using an extension.
+    pub fn from_arrow(name: ComponentName, values: Box<dyn arrow2::Array>) -> Self  { /* ... */ }
+
+    /// Builds an empty `DataCell` from an arrow datatype.
+    //
+    // TODO(cmc): We shouldn't have to specify the component name separately, this should be part of the
+    // metadata by using an extension.
+    pub fn from_datatype(name: ComponentName, datatype: DataType) -> Self  { /* ... */ }
+
+    /// Builds an empty `DataCell` from a component type.
+    //
+    // TODO(cmc): do keep in mind there's a future not too far away where components become a
+    // `(component, type)` tuple kinda thing.
+    pub fn from_component<C: Component>() -> Self  { /* ... */ }
+
+    /// Returns the contents of the cell as an arrow array.
+    pub fn as_arrow(&self) -> Box<dyn arrow2::Array> { /* ... */ }
+
+    /// Returns the contents of the cell as vector of native components.
+    //
+    // TODO(cmc): We could potentially keep the original native component values if the cell was created
+    // using `from_native`.
+    pub fn as_components<C: Component>(&self) -> Vec<C> { /* ... */ }
+}
+
+// TODO(cmc): Some convenient `From` implementations etc
+```
+(The "arrow extension" thing that is mentioned a lot in the comments above is [this](https://docs.rs/arrow2/latest/arrow2/datatypes/enum.DataType.html#variant.Extension).)
+
+`DataRow`, which fills the shoes of today's `MsgBundle`:
+```rust
+/// A row's worth of data, i.e. an event: a list of [`DataCell`]s associated with an auto-generated
+/// [`EventId`], a user-specified [`TimePoint`] and [`EntityPath`], and an expected number of
+/// instances.
+pub struct DataRow {
+    /// Auto-generated [`TUID`], uniquely identifying this event and keeping track of the client's
+    /// wall-clock.
+    event_id: EventId,
+
+    /// User-specified [`TimePoint`] for this event.
+    timepoint: TimePoint,
+
+    /// User-specified [`EntityPath`] for this event.
+    entity_path: EntityPath,
+
+    /// The expected number of values (== component instances) in each cell.
+    ///
+    /// Each cell must have either:
+    /// - 0 instance (clear),
+    /// - 1 instance (splat),
+    /// - `num_instances` instances (standard).
+    num_instances: u32,
+
+    /// The actual cells (== columns).
+    cells: Vec<DataCell>,
```

Comment:
The following list is all the types that are treated separately because they need to be deserialized and stored natively in the `DataStore` since they drive its behavior:
```rust
event_id: Vec<EventId>,
timepoint: Vec<TimePoint>,
entity_path: Vec<EntityPath>,
num_instances: Vec<u32>,
```
None of these are components however, they are merely arrow datatypes.

Everything else is just a component, and as such is passed as a `DataCell`.

Components are completely opaque blobs from the store's PoV; they cannot dictate its behavior since they aren't even deserialized.  
This includes `InstanceKey`s, which are just returned to `re_query` as-is.

The one special thing about instance keys is that they are auto-generated server-side if they are missing; but even then, once they are generated they are still treated as opaque blobs from the store's PoV.

TODO(cmc): add to future work: non-integer instance keys